### PR TITLE
fix(FocusZone): should reset tabindex when focus is outside the zone with prop `shouldResetActiveElementWhenTabFromZone`

### PR DESCRIPTION
**Pull request by YuanboXue-Amber**
For a `FocusZone` with `shouldResetActiveElementWhenTabFromZone` prop specified, when it is updated before receiving focus, the zone does not update tabIndex accordingly.

In this example: https://codesandbox.io/s/fluent-ui-example-forked-hq2lzn
- click on button to add a chat
- tab to focus on button
- tab again, focus should land on the last message. But it in fact land on one message above.
    This is because when one more message is added to `FocusZone`, the zone does not reset tabIndex, while it should

Note that initially I wanted to reset tabIndex on FocusZone re-render. But this is not enough as FocusZone's children can be updated without FocusZone itself re-render. In the end I reset tabIndex every time a keydown is captured, and focus is outside of the zone.

### CODE REVIEW

**Comment by fabricteam** [comment]
## 📊 Bundle size report

<sub>🤖 This report was generated against <a href='https://github.com/microsoft/fluentui/commit/26031bbc5bff98877278557b8d72490751a2fdfe'>26031bbc5bff98877278557b8d72490751a2fdfe</a></sub>
<div id="bundleSizeComment9423"/>

**Comment by fabricteam** [comment]
## Perf Analysis (`@fluentui/react-northstar`)
<details><summary>Perf tests with no regressions</summary>

Scenario | Current PR Ticks | Baseline Ticks | Ratio
:--- | ---:| ---:| ---:
LayoutMinimalPerf.default | [290](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/LayoutMinimalPerf.default.html) | [245](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/LayoutMinimalPerf.default_base.html) | 1.18:1
ProviderMergeThemesPerf.default | [1049](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/ProviderMergeThemesPerf.default.html) | [954](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/ProviderMergeThemesPerf.default_base.html) | 1.1:1
ReactionMinimalPerf.default | [309](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/ReactionMinimalPerf.default.html) | [284](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/ReactionMinimalPerf.default_base.html) | 1.09:1
AttachmentSlotsPerf.default | [950](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/AttachmentSlotsPerf.default.html) | [903](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/AttachmentSlotsPerf.default_base.html) | 1.05:1
ChatDuplicateMessagesPerf.default | [238](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/ChatDuplicateMessagesPerf.default.html) | [227](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/ChatDuplicateMessagesPerf.default_base.html) | 1.05:1
FormMinimalPerf.default | [319](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/FormMinimalPerf.default.html) | [305](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/FormMinimalPerf.default_base.html) | 1.05:1
SkeletonMinimalPerf.default | [287](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/SkeletonMinimalPerf.default.html) | [274](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/SkeletonMinimalPerf.default_base.html) | 1.05:1
ButtonOverridesMissPerf.default | [1247](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/ButtonOverridesMissPerf.default.html) | [1202](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/ButtonOverridesMissPerf.default_base.html) | 1.04:1
LabelMinimalPerf.default | [315](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/LabelMinimalPerf.default.html) | [304](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/LabelMinimalPerf.default_base.html) | 1.04:1
ListMinimalPerf.default | [422](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/ListMinimalPerf.default.html) | [406](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/ListMinimalPerf.default_base.html) | 1.04:1
ButtonSlotsPerf.default | [465](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/ButtonSlotsPerf.default.html) | [453](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/ButtonSlotsPerf.default_base.html) | 1.03:1
DividerMinimalPerf.default | [297](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/DividerMinimalPerf.default.html) | [289](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/DividerMinimalPerf.default_base.html) | 1.03:1
ListWith60ListItems.default | [513](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/ListWith60ListItems.default.html) | [499](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/ListWith60ListItems.default_base.html) | 1.03:1
LoaderMinimalPerf.default | [575](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/LoaderMinimalPerf.default.html) | [560](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/LoaderMinimalPerf.default_base.html) | 1.03:1
RosterPerf.default | [1784](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/RosterPerf.default.html) | [1734](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/RosterPerf.default_base.html) | 1.03:1
PortalMinimalPerf.default | [142](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/PortalMinimalPerf.default.html) | [138](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/PortalMinimalPerf.default_base.html) | 1.03:1
TableManyItemsPerf.default | [1557](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/TableManyItemsPerf.default.html) | [1519](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/TableManyItemsPerf.default_base.html) | 1.03:1
TreeWith60ListItems.default | [134](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/TreeWith60ListItems.default.html) | [130](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/TreeWith60ListItems.default_base.html) | 1.03:1
VideoMinimalPerf.default | [590](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/VideoMinimalPerf.default.html) | [571](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/VideoMinimalPerf.default_base.html) | 1.03:1
AvatarMinimalPerf.default | [156](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/AvatarMinimalPerf.default.html) | [153](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/AvatarMinimalPerf.default_base.html) | 1.02:1
CarouselMinimalPerf.default | [390](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/CarouselMinimalPerf.default.html) | [382](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/CarouselMinimalPerf.default_base.html) | 1.02:1
CheckboxMinimalPerf.default | [2188](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/CheckboxMinimalPerf.default.html) | [2135](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/CheckboxMinimalPerf.default_base.html) | 1.02:1
RadioGroupMinimalPerf.default | [358](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/RadioGroupMinimalPerf.default.html) | [352](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/RadioGroupMinimalPerf.default_base.html) | 1.02:1
IconMinimalPerf.default | [546](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/IconMinimalPerf.default.html) | [535](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/IconMinimalPerf.default_base.html) | 1.02:1
TextAreaMinimalPerf.default | [384](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/TextAreaMinimalPerf.default.html) | [375](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/TextAreaMinimalPerf.default_base.html) | 1.02:1
ButtonMinimalPerf.default | [137](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/ButtonMinimalPerf.default.html) | [136](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/ButtonMinimalPerf.default_base.html) | 1.01:1
HeaderSlotsPerf.default | [632](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/HeaderSlotsPerf.default.html) | [626](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/HeaderSlotsPerf.default_base.html) | 1.01:1
ListCommonPerf.default | [527](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/ListCommonPerf.default.html) | [520](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/ListCommonPerf.default_base.html) | 1.01:1
ListNestedPerf.default | [453](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/ListNestedPerf.default.html) | [448](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/ListNestedPerf.default_base.html) | 1.01:1
MenuButtonMinimalPerf.default | [1398](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/MenuButtonMinimalPerf.default.html) | [1389](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/MenuButtonMinimalPerf.default_base.html) | 1.01:1
ProviderMinimalPerf.default | [333](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/ProviderMinimalPerf.default.html) | [331](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/ProviderMinimalPerf.default_base.html) | 1.01:1
StatusMinimalPerf.default | [554](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/StatusMinimalPerf.default.html) | [551](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/StatusMinimalPerf.default_base.html) | 1.01:1
ToolbarMinimalPerf.default | [744](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/ToolbarMinimalPerf.default.html) | [740](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/ToolbarMinimalPerf.default_base.html) | 1.01:1
TooltipMinimalPerf.default | [1849](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/TooltipMinimalPerf.default.html) | [1830](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/TooltipMinimalPerf.default_base.html) | 1.01:1
CardMinimalPerf.default | [433](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/CardMinimalPerf.default.html) | [432](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/CardMinimalPerf.default_base.html) | 1:1
ChatMinimalPerf.default | [596](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/ChatMinimalPerf.default.html) | [594](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/ChatMinimalPerf.default_base.html) | 1:1
DropdownMinimalPerf.default | [2647](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/DropdownMinimalPerf.default.html) | [2640](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/DropdownMinimalPerf.default_base.html) | 1:1
EmbedMinimalPerf.default | [3453](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/EmbedMinimalPerf.default.html) | [3444](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/EmbedMinimalPerf.default_base.html) | 1:1
FlexMinimalPerf.default | [235](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/FlexMinimalPerf.default.html) | [236](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/FlexMinimalPerf.default_base.html) | 1:1
HeaderMinimalPerf.default | [293](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/HeaderMinimalPerf.default.html) | [292](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/HeaderMinimalPerf.default_base.html) | 1:1
InputMinimalPerf.default | [1076](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/InputMinimalPerf.default.html) | [1073](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/InputMinimalPerf.default_base.html) | 1:1
SplitButtonMinimalPerf.default | [3656](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/SplitButtonMinimalPerf.default.html) | [3643](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/SplitButtonMinimalPerf.default_base.html) | 1:1
CustomToolbarPrototype.default | [2392](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/CustomToolbarPrototype.default.html) | [2388](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/CustomToolbarPrototype.default_base.html) | 1:1
AnimationMinimalPerf.default | [441](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/AnimationMinimalPerf.default.html) | [444](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/AnimationMinimalPerf.default_base.html) | 0.99:1
AttachmentMinimalPerf.default | [120](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/AttachmentMinimalPerf.default.html) | [121](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/AttachmentMinimalPerf.default_base.html) | 0.99:1
ChatWithPopoverPerf.default | [314](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/ChatWithPopoverPerf.default.html) | [316](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/ChatWithPopoverPerf.default_base.html) | 0.99:1
DatepickerMinimalPerf.default | [4592](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/DatepickerMinimalPerf.default.html) | [4651](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/DatepickerMinimalPerf.default_base.html) | 0.99:1
GridMinimalPerf.default | [267](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/GridMinimalPerf.default.html) | [269](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/GridMinimalPerf.default_base.html) | 0.99:1
MenuMinimalPerf.default | [690](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/MenuMinimalPerf.default.html) | [695](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/MenuMinimalPerf.default_base.html) | 0.99:1
PopupMinimalPerf.default | [514](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/PopupMinimalPerf.default.html) | [517](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/PopupMinimalPerf.default_base.html) | 0.99:1
RefMinimalPerf.default | [176](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/RefMinimalPerf.default.html) | [178](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/RefMinimalPerf.default_base.html) | 0.99:1
SliderMinimalPerf.default | [1395](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/SliderMinimalPerf.default.html) | [1405](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/SliderMinimalPerf.default_base.html) | 0.99:1
TextMinimalPerf.default | [272](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/TextMinimalPerf.default.html) | [276](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/TextMinimalPerf.default_base.html) | 0.99:1
AccordionMinimalPerf.default | [116](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/AccordionMinimalPerf.default.html) | [118](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/AccordionMinimalPerf.default_base.html) | 0.98:1
BoxMinimalPerf.default | [270](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/BoxMinimalPerf.default.html) | [275](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/BoxMinimalPerf.default_base.html) | 0.98:1
ImageMinimalPerf.default | [316](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/ImageMinimalPerf.default.html) | [324](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/ImageMinimalPerf.default_base.html) | 0.98:1
ItemLayoutMinimalPerf.default | [979](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/ItemLayoutMinimalPerf.default.html) | [1000](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/ItemLayoutMinimalPerf.default_base.html) | 0.98:1
DropdownManyItemsPerf.default | [531](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/DropdownManyItemsPerf.default.html) | [552](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/DropdownManyItemsPerf.default_base.html) | 0.96:1
TreeMinimalPerf.default | [628](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/TreeMinimalPerf.default.html) | [651](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/TreeMinimalPerf.default_base.html) | 0.96:1
AlertMinimalPerf.default | [215](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/AlertMinimalPerf.default.html) | [228](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/AlertMinimalPerf.default_base.html) | 0.94:1
SegmentMinimalPerf.default | [256](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/SegmentMinimalPerf.default.html) | [275](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/SegmentMinimalPerf.default_base.html) | 0.93:1
TableMinimalPerf.default | [298](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/TableMinimalPerf.default.html) | [321](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/TableMinimalPerf.default_base.html) | 0.93:1
DialogMinimalPerf.default | [570](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/DialogMinimalPerf.default.html) | [626](https://fluentuipr.z22.web.core.windows.net/pull/24463/perf-test-northstar/DialogMinimalPerf.default_base.html) | 0.91:1

</details><div id="perfComment9424"/>

File: packages/fluentui/react-bindings/test/FocusZone/FocusZone-test.tsx [hunk 1]
```diff
@@ -1852,4 +1852,66 @@ describe('FocusZone', () => {
     expect(buttonB.tabIndex).toBe(-1);
     expect(buttonC.tabIndex).toBe(0);
   });
+
+  it('Update tabIndex when FocusZone is changed before focused', () => {
```
[the remark this answers is not in the document]
Comment (YuanboXue-Amber) [reply packages/fluentui/react-bindings/test/FocusZone/FocusZone-test.tsx:1856]: why no sense when the test works as expected? I was initially thinking cypress would be more readable; but eventually did this test as we have all FTZ test here.

**Review by ling1726 — APPROVED** [review]
Honestly this seems like a userland problem if we could expose an imperative ref like we do with popper

```ts
const fzRef = React.useRef<FocusZomeImperativeRef>();

fzRef.updateTabIndexes();
```

...but that opens a whole other can of worms. I couldn't think of a better solution so here is your approval :)

File: packages/fluentui/react-bindings/test/FocusZone/FocusZone-test.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (ling1726) [reply packages/fluentui/react-bindings/test/FocusZone/FocusZone-test.tsx:1856]: Honestly I think most of our tests should be run on browser, they are quite uncomplicated and give us the 'closest' realistic scenario.

Especially for something like focus, it's really browser specific. It's not a blocker but setting up some FZ tests for the browser would be nice

**Pull request merged**